File: python/navigation/global_planning/sampling/eit_star.py
"""EIT* — Effort Informed Trees (Strub & Gammell 2022, IJRR).

EIT* extends the AIT* idea — a reverse search over the random geometric graph
(RGG) produces an adaptive cost-to-go heuristic that a forward best-first search
consumes, with edges found in collision fed back so the heuristic adapts — by
*also* estimating the **validation effort** of the remaining path. Among
candidates of near-equal cost the forward search then prefers the ones that are
cheaper to collision-check (fewer / shorter sub-segments), surfacing feasible
solutions sooner. Strub & Gammell (2022) motivate this by the observation that
in high-dimensional or expensive-to-validate spaces the collision checker, not
the graph search, dominates runtime.

What this module implements (the faithful core):
- Per batch, grow the RGG exactly as AIT*/BIT*: draw ``batch_size`` informed
  samples (Gammell et al. 2014 ellipse), build the within-radius neighbour graph,
  and filter it against a set of edges already found in collision that persists
  and only grows across batches (the adaptive edge-invalidation feedback of the
  AIT* lineage).
- Reverse search: two independent single-criterion Dijkstra passes from the goal
  over the filtered graph — ``h_hat`` (cost-to-go, edge weight ``distance``) and
  ``e_hat`` (effort-to-go, edge weight ``effort``).
- Forward search: lazy-deletion best-first keyed by the lexicographic pair
  ``(g + h_hat, effort_g + e_hat)`` — cost primary, effort the tie-break, exactly
  the "solutions of equal cost differentiated by effort" ordering of the paper.
  Motions are collision-checked lazily on dequeue; an invalid motion is recorded
  in the persistent invalid-edge set and skipped.

Per-edge effort needs no new capability: ``effort(u, v)`` = the number of
``step_size``-sized sub-segments a discretized motion validator would check =
``max(1, round(distance(u, v) / step_size))``. This is a proxy for collision-check
cost that reads only the ``SamplingSpace`` distance, never map internals.

Explicit simplifications vs. the full paper (Strub & Gammell 2022, IJRR):
- The reverse search is recomputed from scratch each batch over the accumulated
  sample array rather than repaired incrementally (LPA*-style) as in the paper;
  only ``c_best`` and the invalid-edge set carry across batches.
- Cost and effort heuristics come from two clean independent single-criterion
  Dijkstra passes rather than the paper's more integrated joint treatment.
- Effort is the simple ``distance / step_size`` discretization proxy, not a
  learned or measured validator-cost model.

Properties: probabilistically complete, almost-surely asymptotically optimal,
anytime; additionally biases toward cheaper-to-validate solutions among
near-equal-cost candidates.
"""

from __future__ import annotations

import heapq
import time

import numpy as np

from navigation.core.capabilities import Capability, SamplingSpace
from navigation.core.planner import GlobalPlanner
from navigation.core.trace import TraceRecorder
from navigation.core.types import PlanResult, PlanStats, Point

from ._sampling import informed_sample, path_length, radius_neighbors, rgg_radius

_INF = float("inf")

# Undirected edge key: order-independent so (u,v) and (v,u) hit the same entry in
# the persistent invalid-edge set.
_Edge = tuple[int, int]


def _edge_key(a: int, b: int) -> _Edge:
    return (a, b) if a < b else (b, a)


def _dijkstra_from(
    source: int,
    adjacency: list[list[int]],
    weight: list[list[float]],
) -> list[float]:
    """Single-source shortest cost over a fixed undirected graph.

    Runs one clean single-criterion relaxation; EIT* invokes it twice from the
    goal (distance weights, then effort weights) to build the two reverse
    heuristics independently (Strub & Gammell 2022).
    """
    n = len(adjacency)
    dist = [_INF] * n
    dist[source] = 0.0
    heap: list[tuple[float, int]] = [(0.0, source)]
    while heap:
        d, u = heapq.heappop(heap)
        if d > dist[u]:
            continue
        for k, v in enumerate(adjacency[u]):
            nd = d + weight[u][k]
            if nd < dist[v]:
                dist[v] = nd
                heapq.heappush(heap, (nd, v))
    return dist


class EITStar(GlobalPlanner[Point, "SamplingSpace[Point]"]):
    @property
    def name(self) -> str:
        return "eit_star"

    def required_capabilities(self) -> set[Capability]:
        return {Capability.SAMPLING_SPACE}

    def plan(
        self,
        space: SamplingSpace[Point],
        start: Point,
        goal: Point,
        recorder: TraceRecorder | None = None,
    ) -> PlanResult[Point]:
        t0 = time.monotonic()
        batch_size = self.params.get_int("batch_size")
        max_batches = self.params.get_int("max_batches")
        gamma = self.params.get_float("gamma")
        step_size = self.params.get_float("step_size")
        rng = np.random.default_rng(self.params.get_int("seed"))

        # Sample array grows across batches: index 0 = start, 1 = goal (permanent).
        points: list[Point] = [start, goal]
        start_idx, goal_idx = 0, 1
        invalid_edges: set[_Edge] = set()
        c_best = _INF
        expanded = 0
        # Forward-search cost/effort/parent from the final batch — the incumbent is
        # read off these once batches are exhausted (batch-recomputed, per the
        # documented simplification of the AIT*/EIT* reverse search).
        g: list[float] = [_INF, _INF]
        parent: list[int] = [-1, -1]

        for _ in range(max_batches):
            g, parent, c_best, expanded = self._run_batch(
                space, start, goal, start_idx, goal_idx, batch_size, gamma,
                step_size, rng, points, invalid_edges, c_best, expanded, recorder,
            )

        # --- extract incumbent from the final batch's forward tree ----------------
        success = g[goal_idx] < _INF
        path: list[Point] = []
        if success:
            node = goal_idx
            while node != -1:
                path.append(points[node])
                node = parent[node]
            path.reverse()
        cost = path_length(space, path) if success else 0.0
        if success and recorder is not None:
            recorder.path_found(path)

        runtime = time.monotonic() - t0
        n = len(points)
        if recorder is not None:
            recorder.planning_finished(
                success,
                {
                    "runtime_sec": runtime,
                    "path_cost": cost,
                    "expanded_nodes": float(expanded),
                    "samples": float(n),
                    "tree_size": float(n),
                },
            )
        stats = PlanStats(expanded_nodes=expanded, samples=n, tree_size=n, iterations=expanded)
        return PlanResult(success, path, cost, stats)

    def _run_batch(
        self,
        space: SamplingSpace[Point],
        start: Point,
        goal: Point,
        start_idx: int,
        goal_idx: int,
        batch_size: int,
        gamma: float,
        step_size: float,
        rng: np.random.Generator,
        points: list[Point],
        invalid_edges: set[_Edge],
        c_best: float,
        expanded: int,
        recorder: TraceRecorder | None,
    ) -> tuple[list[float], list[int], float, int]:
        # --- draw a new batch (informed once a solution exists) -------------------
        drawn = 0
        for _ in range(batch_size * 40):
            if drawn >= batch_size:
                break
            q = informed_sample(space, start, goal, c_best, rng)
            if not space.is_state_valid(q):
                continue
            points.append(q)
            drawn += 1
            if recorder is not None:
                recorder.sample_drawn(q)

        n = len(points)
        radius = rgg_radius(gamma, n)
        nbr = radius_neighbors(space, points, radius)

        def effort(a: int, b: int) -> int:
            # Number of step_size-sized sub-segments a discretized validator would
            # check on edge (a,b): a capability-free proxy for collision-check cost
            # (Strub & Gammell 2022).
            return max(1, round(space.distance(points[a], points[b]) / step_size))

        # --- filtered adjacency + parallel distance/effort weights ----------------
        adjacency: list[list[int]] = [[] for _ in range(n)]
        dist_w: list[list[float]] = [[] for _ in range(n)]
        eff_w: list[list[float]] = [[] for _ in range(n)]
        for u in range(n):
            for v in nbr[u]:
                if _edge_key(u, v) in invalid_edges:
                    continue
                adjacency[u].append(v)
                dist_w[u].append(space.distance(points[u], points[v]))
                eff_w[u].append(float(effort(u, v)))

        # --- reverse search: two independent Dijkstra passes from the goal --------
        h_hat = _dijkstra_from(goal_idx, adjacency, dist_w)
        e_hat = _dijkstra_from(goal_idx, adjacency, eff_w)

        # --- forward search: lazy-deletion best-first over (cost, effort) ---------
        g = [_INF] * n
        effort_g = [_INF] * n
        parent = [-1] * n
        closed = [False] * n
        g[start_idx] = 0.0
        effort_g[start_idx] = 0.0
        # Heap key: (g + h_hat, effort_g + e_hat, idx). Python tuples compare
        # lexicographically, so cost is primary and effort the tie-break exactly as
        # the paper intends; idx makes the order deterministic.
        heap: list[tuple[float, float, int]] = [(h_hat[start_idx], e_hat[start_idx], start_idx)]

        while heap:
            _, _, v = heapq.heappop(heap)
            if closed[v]:
                continue
            closed[v] = True
            expanded += 1
            for x in adjacency[v]:
                d = space.distance(points[v], points[x])
                if recorder is not None:
                    # Cost is the primary reported metric for a candidate.
                    recorder.candidate_evaluated(points[x], g[v] + d)
                if not space.is_motion_valid(points[v], points[x]):
                    invalid_edges.add(_edge_key(v, x))
                    continue
                new_g = g[v] + d
                new_effort = effort_g[v] + float(effort(v, x))
                # Lexicographic acceptance: cost primary, cumulative effort the
                # tie-break, consistent with the priority-queue ordering.
                if (new_g, new_effort) < (g[x], effort_g[x]):
                    first = parent[x] == -1
                    g[x] = new_g
                    effort_g[x] = new_effort
                    parent[x] = v
                    closed[x] = False  # improved: allow re-expansion this batch
                    heapq.heappush(heap, (new_g + h_hat[x], new_effort + e_hat[x], x))
                    if recorder is not None:
                        if first:
                            recorder.edge_added(points[x], points[v], d)
                        else:
                            recorder.rewire(points[x], points[v])
                    if x == goal_idx and new_g < c_best:
                        c_best = new_g
                        if recorder is not None:
                            recorder.path_found(_reconstruct(points, parent, goal_idx))

        return g, parent, c_best, expanded


def _reconstruct(points: list[Point], parent: list[int], goal_idx: int) -> list[Point]:
    path: list[Point] = []
    node = goal_idx
    while node != -1:
        path.append(points[node])
        node = parent[node]
    path.reverse()
    return path

File: cpp/src/global_planning/sampling/eit_star.cpp
#include "navigation/global_planning/sampling/eit_star.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <functional>
#include <limits>
#include <queue>
#include <set>
#include <tuple>
#include <utility>
#include <vector>

#include "navigation/global_planning/sampling/sampling_common.hpp"

namespace navigation::global_planning {

namespace {
constexpr double kInf = std::numeric_limits<double>::infinity();

using Edge = std::pair<int, int>;

// Undirected edge key: order-independent so (u,v) and (v,u) share one entry in
// the persistent invalid-edge set.
Edge edge_key(int a, int b) { return a < b ? Edge{a, b} : Edge{b, a}; }

// Single-source shortest cost over a fixed undirected graph. EIT* invokes it
// twice from the goal (distance weights, then effort weights) to build the two
// reverse heuristics independently (Strub & Gammell 2022).
std::vector<double> dijkstra_from(int source, const std::vector<std::vector<int>>& adjacency,
                                  const std::vector<std::vector<double>>& weight) {
  int n = static_cast<int>(adjacency.size());
  std::vector<double> dist(static_cast<size_t>(n), kInf);
  dist[static_cast<size_t>(source)] = 0.0;
  using Entry = std::pair<double, int>;  // (cost, idx)
  std::priority_queue<Entry, std::vector<Entry>, std::greater<Entry>> heap;
  heap.emplace(0.0, source);
  while (!heap.empty()) {
    auto [d, u] = heap.top();
    heap.pop();
    if (d > dist[static_cast<size_t>(u)]) continue;
    const auto& nbrs = adjacency[static_cast<size_t>(u)];
    const auto& wts = weight[static_cast<size_t>(u)];
    for (size_t k = 0; k < nbrs.size(); ++k) {
      int v = nbrs[k];
      double nd = d + wts[k];
      if (nd < dist[static_cast<size_t>(v)]) {
        dist[static_cast<size_t>(v)] = nd;
        heap.emplace(nd, v);
      }
    }
  }
  return dist;
}
}  // namespace

core::PlanResult<Point> EitStarPlanner::plan(SamplingSpace<Point>& space, const Point& start,
                                             const Point& goal, TraceRecorder* recorder) {
  const int batch_size = params_.get_int("batch_size");
  const int max_batches = params_.get_int("max_batches");
  const double gamma = params_.get_float("gamma");
  const double step_size = params_.get_float("step_size");
  std::mt19937 rng(static_cast<unsigned>(params_.get_int("seed")));

  auto t0 = std::chrono::steady_clock::now();
  // Sample array grows across batches: 0 = start, 1 = goal (permanent).
  std::vector<Point> points{start, goal};
  const int start_idx = 0, goal_idx = 1;
  std::set<Edge> invalid_edges;
  double c_best = kInf;
  int expanded = 0;
  // Forward-search cost/parent from the final batch — the incumbent is read off
  // these once batches are exhausted (batch-recomputed reverse/forward search, per
  // the documented simplification of the AIT*/EIT* lineage).
  std::vector<double> g{kInf, kInf};
  std::vector<int> parent{-1, -1};

  // Number of step_size-sized sub-segments a discretized validator would check on
  // edge (a,b): a capability-free proxy for collision-check cost (Strub & Gammell 2022).
  auto effort = [&](int a, int b) -> double {
    double d = space.distance(points[static_cast<size_t>(a)], points[static_cast<size_t>(b)]);
    return static_cast<double>(std::max(1L, std::lround(d / step_size)));
  };

  auto reconstruct = [&](const std::vector<int>& par) {
    std::vector<Point> path;
    for (int node = goal_idx; node != -1; node = par[static_cast<size_t>(node)]) {
      path.push_back(points[static_cast<size_t>(node)]);
    }
    std::reverse(path.begin(), path.end());
    return path;
  };

  for (int batch = 0; batch < max_batches; ++batch) {
    // --- draw a new batch (informed once a solution exists) --------------------
    int drawn = 0;
    for (int attempt = 0; attempt < batch_size * 40 && drawn < batch_size; ++attempt) {
      Point q = informed_sample(space, start, goal, c_best, rng);
      if (!space.is_state_valid(q)) continue;
      points.push_back(q);
      ++drawn;
      if (recorder) recorder->sample_drawn(q);
    }

    int n = static_cast<int>(points.size());
    double radius = rgg_radius(gamma, n);
    std::vector<std::vector<int>> nbr = radius_neighbors(space, points, radius);

    // --- filtered adjacency + parallel distance/effort weights -----------------
    std::vector<std::vector<int>> adjacency(static_cast<size_t>(n));
    std::vector<std::vector<double>> dist_w(static_cast<size_t>(n));
    std::vector<std::vector<double>> eff_w(static_cast<size_t>(n));
    for (int u = 0; u < n; ++u) {
      for (int v : nbr[static_cast<size_t>(u)]) {
        if (invalid_edges.count(edge_key(u, v))) continue;
        adjacency[static_cast<size_t>(u)].push_back(v);
        dist_w[static_cast<size_t>(u)].push_back(
            space.distance(points[static_cast<size_t>(u)], points[static_cast<size_t>(v)]));
        eff_w[static_cast<size_t>(u)].push_back(effort(u, v));
      }
    }

    // --- reverse search: two independent Dijkstra passes from the goal ---------
    std::vector<double> h_hat = dijkstra_from(goal_idx, adjacency, dist_w);
    std::vector<double> e_hat = dijkstra_from(goal_idx, adjacency, eff_w);

    // --- forward search: lazy-deletion best-first over (cost, effort) ----------
    g.assign(static_cast<size_t>(n), kInf);
    std::vector<double> effort_g(static_cast<size_t>(n), kInf);
    parent.assign(static_cast<size_t>(n), -1);
    std::vector<char> closed(static_cast<size_t>(n), 0);
    g[static_cast<size_t>(start_idx)] = 0.0;
    effort_g[static_cast<size_t>(start_idx)] = 0.0;
    // Key (g + h_hat, effort_g + e_hat, idx): std::tuple compares lexicographically,
    // so cost is primary and effort the tie-break exactly as the paper intends; idx
    // makes the order deterministic.
    using FEntry = std::tuple<double, double, int>;
    std::priority_queue<FEntry, std::vector<FEntry>, std::greater<FEntry>> heap;
    heap.emplace(h_hat[static_cast<size_t>(start_idx)], e_hat[static_cast<size_t>(start_idx)],
                 start_idx);

    while (!heap.empty()) {
      int v = std::get<2>(heap.top());
      heap.pop();
      if (closed[static_cast<size_t>(v)]) continue;
      closed[static_cast<size_t>(v)] = 1;
      ++expanded;
      for (int x : adjacency[static_cast<size_t>(v)]) {
        double d = space.distance(points[static_cast<size_t>(v)], points[static_cast<size_t>(x)]);
        if (!space.is_motion_valid(points[static_cast<size_t>(v)],
                                   points[static_cast<size_t>(x)])) {
          invalid_edges.insert(edge_key(v, x));
          continue;
        }
        double new_g = g[static_cast<size_t>(v)] + d;
        double new_effort = effort_g[static_cast<size_t>(v)] + effort(v, x);
        // Lexicographic acceptance: cost primary, cumulative effort the tie-break,
        // consistent with the priority-queue ordering.
        if (std::make_pair(new_g, new_effort) <
            std::make_pair(g[static_cast<size_t>(x)], effort_g[static_cast<size_t>(x)])) {
          bool first = parent[static_cast<size_t>(x)] == -1;
          g[static_cast<size_t>(x)] = new_g;
          effort_g[static_cast<size_t>(x)] = new_effort;
          parent[static_cast<size_t>(x)] = v;
          closed[static_cast<size_t>(x)] = 0;  // improved: allow re-expansion this batch
          heap.emplace(new_g + h_hat[static_cast<size_t>(x)],
                       new_effort + e_hat[static_cast<size_t>(x)], x);
          if (recorder) {
            // Emit candidate_evaluated only for feasible, improving edges (not
            // every relaxed neighbour) so the trace stays renderable by replay.py
            // — matches BIT*'s emit-on-accept scale rather than exploding on the
            // batch-recomputed graph. Cost is the primary reported metric.
            recorder->candidate_evaluated(points[static_cast<size_t>(x)], new_g);
            if (first)
              recorder->edge_added(points[static_cast<size_t>(x)], points[static_cast<size_t>(v)],
                                   d);
            else
              recorder->rewire(points[static_cast<size_t>(x)], points[static_cast<size_t>(v)]);
          }
          if (x == goal_idx && new_g < c_best) {
            c_best = new_g;
            if (recorder) recorder->path_found(reconstruct(parent));
          }
        }
      }
    }
  }

  // --- extract incumbent from the final batch's forward tree -----------------
  bool success = g[static_cast<size_t>(goal_idx)] < kInf;
  std::vector<Point> path;
  if (success) path = reconstruct(parent);
  double cost = success ? path_length(space, path) : 0.0;
  if (success && recorder) recorder->path_found(path);

  double rt = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
  int n = static_cast<int>(points.size());
  emit_finished_batch(recorder, success, cost, expanded, n, n, rt);
  core::PlanResult<Point> result;
  result.success = success;
  result.path = std::move(path);
  result.cost = cost;
  result.stats = {expanded, n, expanded, n};
  return result;
}

}  // namespace navigation::global_planning
